### PR TITLE
`[Move]` Part-6 Classes into Different Library - `Neo.Extensions`

### PR DESCRIPTION
**Pull request by cschuchardt88**
# Description
Reorganized and move classes from `neo` library to `neo.extensions` library. 

### Fixed problems and warnings that occurred. Along with naming violations.

# Change Log
 - Moved methods from `Neo.IO.Helper` class
 - Added `BinaryReaderExtensions`, `BinaryWriterExtensions`, `MemoryReaderExtensions`, `SpanExtensions` classes
 - Added more methods to `StringExtensions`, `ByteExtensions` `MemoryExtensions`
 - Added `using` statement
 - Fixed unit tests

## Type of change

- [x] Optimization (the change is only an optimization)
- [x] Style (the change is only a code style for better maintenance or standard purpose)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# How Has This Been Tested?
- Unit Testing

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: .github/workflows/docker.yml [hunk 1]
```diff
@@ -0,0 +1,77 @@
+name: Publish (docker-image)
```
Comment (Jim8y) [inline .github/workflows/docker.yml]: why docker change in this pr?

Comment (cschuchardt88) [reply .github/workflows/docker.yml]: I guess it was a github glitch, its looks like it corrected itself.

File: tests/Neo.Extensions.Tests/UT_BigIntegerExtensions.cs [hunk 2]
```diff
@@ -21,6 +21,7 @@ namespace Neo.Extensions.Tests
     [TestClass]
     public class UT_BigIntegerExtensions
     {
+
```
Comment (shargon) [inline tests/Neo.Extensions.Tests/UT_BigIntegerExtensions.cs:24]:
```suggestion
```

File: tests/Neo.Extensions.Tests/UT_StringExtensions.cs [hunk 3]
```diff
@@ -42,5 +42,157 @@ public void TestGetVarSizeString()
...
+                    Assert.AreEqual(3, result);
+                }
+                else
+                {
+                    int result = UnsafeData.GetVarSize(0xFFFFFF);
+                    Assert.AreEqual(5, result);
+                }
+            }
+        }
+
```
Comment (shargon) [inline tests/Neo.Extensions.Tests/UT_StringExtensions.cs:68]:
```suggestion
```

File: tests/Neo.Extensions.Tests/UT_StringExtensions.cs [hunk 4]
```diff
@@ -42,5 +42,156 @@ public void TestGetVarSizeString()
             int result = "AA".GetVarSize();
             Assert.AreEqual(3, result);
         }
+
+        [TestMethod]
+        public void TestGetVarSizeInt()
+        {
+            for (int i = 0; i < 3; i++)
```
Comment (shargon) [inline tests/Neo.Extensions.Tests/UT_StringExtensions.cs:49]: Why a for ?

File: tests/Neo.Extensions.Tests/UT_StringExtensions.cs [hunk 5]
```diff
@@ -42,5 +42,156 @@ public void TestGetVarSizeString()
...
+                    int result = UnsafeData.GetVarSize(0xFFFFFF);
+                    Assert.AreEqual(5, result);
+                }
+            }
+        }
+
+        [TestMethod]
+        public void TestGetVarSizeGeneric()
+        {
+            for (int i = 0; i < 9; i++)
```
Comment (shargon) [inline tests/Neo.Extensions.Tests/UT_StringExtensions.cs:72]: ?? why a for if i is not used in the test?

**Review by shargon — CHANGES_REQUESTED** [review]

File: tests/Neo.Extensions.Tests/UT_StringExtensions.cs (hunk 4 above)
Comment (cschuchardt88) [reply tests/Neo.Extensions.Tests/UT_StringExtensions.cs:49]: not sure, i'm just moving the tests.

File: tests/Neo.Extensions.Tests/UT_StringExtensions.cs (hunk 5 above)
Comment (cschuchardt88) [reply tests/Neo.Extensions.Tests/UT_StringExtensions.cs:72]: not sure, i'm just moving the tests.

File: src/Neo/Extensions/SpanExtensions.cs [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+                throw new FormatException();
+            return result;
+        }
+
+        /// <summary>
+        /// Decompresses the specified data using the LZ4 algorithm.
+        /// </summary>
+        /// <param name="data">The compressed data.</param>
+        /// <param name="maxOutput">The maximum data size after decompression.</param>
+        /// <returns>The original data.</returns>
```
Comment (Jim8y) [inline src/Neo/Extensions/SpanExtensions.cs:69]: I know that adding new methods will make it easier to do the work, and actually it wont cause any trouble, but please avoid doing so.

**Review by Jim8y — APPROVED** [review]

**Review by vncoelho — APPROVED** [review]

File: src/Neo/Extensions/SpanExtensions.cs (hunk 6 above)
Comment (shargon) [reply src/Neo/Extensions/SpanExtensions.cs:69]: Agree, please don't add if the PR is so big and is not related with this

**Review by shargon — COMMENTED** [review]
@superboyiii could you check if there are any storage change in both networks?

**Comment by superboyiii** [comment]
> @superboyiii could you check if there are any storage change in both networks?

Sure

**Comment by superboyiii** [comment]
Mainnet test result shows it's compatbile. I will try testnet as well.

**Comment by superboyiii** [comment]
Data is compatible on testnet as well.

**Review by shargon — APPROVED** [review]

**Pull request merged**